File: src/api/dotnet/ConstructorList.cs
/*++
Copyright (c) 2012 Microsoft Corporation

Module Name:

    ConstructorList.cs

Abstract:

    Z3 Managed API: Constructor Lists

Author:

    Christoph Wintersteiger (cwinter) 2012-11-23

Notes:
    
--*/

using System.Diagnostics;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;


namespace Microsoft.Z3
{
    /// <summary>
    /// Lists of constructors
    /// </summary>
    public class ConstructorList : Z3Object
    {
        /// <summary>
        /// Destructor.
        /// </summary>
        ~ConstructorList()
        {
            Native.Z3_del_constructor_list(Context.nCtx, NativeObject);
        }

        #region Internal
        internal ConstructorList(Context ctx, IntPtr obj)
            : base(ctx, obj)
        {
            Debug.Assert(ctx != null);
        }

        internal ConstructorList(Context ctx, Constructor[] constructors)
            : base(ctx)
        {
            Debug.Assert(ctx != null);
            Debug.Assert(constructors != null);

            NativeObject = Native.Z3_mk_constructor_list(Context.nCtx, (uint)constructors.Length, Constructor.ArrayToNative(constructors));
        }
        #endregion
    }
}

File: src/api/java/ConstructorList.java
/**
Copyright (c) 2012-2014 Microsoft Corporation
   
Module Name:

    ConstructorList.java

Abstract:

Author:

    @author Christoph Wintersteiger (cwinter) 2012-03-15

Notes:
    
**/

package com.microsoft.z3;

import java.lang.ref.ReferenceQueue;

/**
 * Lists of constructors
 **/
public class ConstructorList<R> extends Z3Object {

    ConstructorList(Context ctx, long obj)
    {
        super(ctx, obj);
    }

    @Override
    void incRef() {
        // Constructor lists are not reference counted.
    }

    @Override
    void addToReferenceQueue() {
        getContext().getReferenceQueue().storeReference(this, ConstructorListRef::new);
    }

    ConstructorList(Context ctx, Constructor<R>[] constructors)
    {
        super(ctx, Native.mkConstructorList(ctx.nCtx(),
                constructors.length,
                Constructor.arrayToNative(constructors)));
    }

    private static class ConstructorListRef extends Z3ReferenceQueue.Reference<ConstructorList<?>> {

        private ConstructorListRef(ConstructorList<?> referent, ReferenceQueue<Z3Object> q) {
            super(referent, q);
        }

        @Override
        void decRef(Context ctx, long z3Obj) {
            Native.delConstructorList(ctx.nCtx(), z3Obj);
        }
    }
}
